test/utilities.jl

```diff
@@ -0,0 +1,108 @@
+shape, the locations of the max errors are also printed.
+
+# Keywords
+- `rtol`: The desired relative tolerance
+- `atol`: The desired absolute tolerance
+- `err_msg`: An optional error message that can be printed when the arrays are not close
+- `error_location_fmt`: An optional keyword used to specify a function that creates a string 
+  describing the location in the arrays where values are unequal, given a `CartesianIndex` object.
+  This should not be specified in cases where actual and reference have different shapes.
+"""
+function assert_allclose(actual, reference; rtol = 1e-7, atol = 0.0, err_msg = nothing,
+                         error_location_fmt = nothing)
+    # make the initial check fast!
+    if all( abs.(actual .- reference) .<= (rtol .* abs.(reference) .+ atol))
+        return true
+    end
+
+    common_layout = size(actual) == size(reference)
+    if isnothing(error_location_fmt)
+        error_location_fmt = (arg) -> string(arg.I)
+    elseif !common_layout
+        throw("error_location_fmt must be nothing b/c actual and reference have different shapes")
+    end
+
+    # now let's provide a detailed error message (this can take longer):
+
+    # first, let's determine what the max error is (and possibly where it happended)
+    diff = abs.(actual .- reference)
+    max_abs_diff_loc = argmax(diff)
+    max_abs_diff = diff[max_abs_diff_loc]
+
+    relative_diff = diff ./ abs.(reference)
+    max_rel_error_magnitude_loc = argmax(relative_diff)
+    max_rel_error_magnitude = relative_diff[max_rel_error_magnitude_loc]
+
+    lines = isnothing(err_msg) ? [] : [err_msg]
+    push!(lines, string("The arrays aren't consistent to within atol = ", atol,
+                        " rtol = ", rtol))
+
+    if (max_rel_error_magnitude == -1.0) && ((rtol != 0.0) || (atol == 0.0))
+        push!(lines, "Max Rel Diff Magnitude: ∞")
+    elseif ((rtol != 0.0) || (atol == 0.0)) && isnothing(max_rel_error_magnitude_loc)
+        push!(lines, string("Max Rel Diff Magnitude: ", max_rel_error_magnitude))
+    elseif ((rtol != 0.0) || (atol == 0.0))
+        push!(lines, string("Max Rel Diff Magnitude: ", max_rel_error_magnitude, " at ",
+                            error_location_fmt(max_rel_error_magnitude_loc)))
+    end
+
+    if ((atol != 0.0) || (rtol == 0.0)) && isnothing(max_abs_diff_loc)
+        push!(lines, string("Max Abs Diff Magnitude: ", max_abs_diff))
+    elseif ((atol != 0.0) || (rtol == 0.0))
+        push!(lines, string("Max Abs Diff Magnitude: ", max_abs_diff, " at ",
+                            error_location_fmt( max_abs_diff_loc)))
+    end
+
+    throw(ErrorException(join(lines, "\n")))
+    false
+end
+
+"""
+assert_allclose_grid(actual, reference, independent_vars; [rtol], [atol], [err_msg])
+
+This function is a convenience wrapper around assert_allclose that provides slightly nicer error 
+messages when actual and reference have the same shape. 
```

Comment:
Feel free to reword, but I think it's worth explaining this a bit more.

```suggestion
This function is a convenience wrapper around assert_allclose that provides slightly nicer error 
messages based on the independent variables generating the grid.  It requires that actual and
reference have the same shape. 
```